### PR TITLE
arrows: add ability to change label placement

### PR DESCRIPTION
**Pull request by mimecuvalo**
This adds the ability to drag the label on an arrow to a different location within the line segment/arc.

https://github.com/tldraw/tldraw/assets/469604/dbd2ee35-bebc-48d6-b8ee-fcf12ce91fa5

- A lot of the complexity lay in ensuring a fixed distance from the ends of the arrowheads.
- I added a new type of handle `text-adjust` that makes the text box the very handle itself.
- I added a `ARROW_HANDLES` enum - we should use more enums!
- The bulk of the changes are in ArrowShapeUtil — check that out in particular obviously :)

Along the way, I tried to improve a couple spots as I touched them:
- added some more documentation to Vec.ts because some of the functions in there were obscure/new to me. (at least the naming, hah)
- added `getPointOnCircle` which was being done in a couple places independently and refactored those places.

### Questions
- the `getPointOnCircle` API changed. Is this considered breaking and/or should I leave the signature the same? Wasn't sure if it was a big deal or not.
- I made `labelPosition` in the schema always but I guess it could have been optional? Lemme know if there's a preference.
- Any feedback on tests? Happy to expand those if necessary.

### Change Type

- [ ] `patch` — Bug fix
- [x] `minor` — New feature
- [ ] `major` — Breaking change
- [ ] `dependencies` — Changes to package dependencies[^1]
- [ ] `documentation` — Changes to the documentation only[^2]
- [ ] `tests` — Changes to any test code only[^2]
- [ ] `internal` — Any other changes that don't affect the published package[^2]
- [ ] I don't know

[^1]: publishes a `patch` release, for devDependencies use `internal`
[^2]: will not publish a new version

### Test Plan

1. For arrow in [straightArrow, curvedArrow] test the following:
   a. Label in the middle
   b. Label at both ends of the arrow
   c. Test arrows in different directions
   d. Rotating the endpoints and seeing that the label stays at the end of the arrow at a fixed width.
   e. Test different stroke widths.
   f. Test with different arrowheads.
2. Also, test arcs that are more circle like than arc-like.

- [x] Unit Tests
- [ ] End to end tests

### Release Notes

- Adds ability to change label position on arrows.

### CODE REVIEW

**Review by ds300 — COMMENTED** [review]
This is looking great! :tada:


Kapture 2024-01-19 at 14 55 49

As already mentioned in person, I find this interaction quite problematic:

- start with no shapes selected
- decide to change the position of a arrow label
- click on it to drag
- the whole arrow moves, not just the label

My ideal interaction for this would involve the label and the arrow having independent selections states. This could be achieved with the upcoming bindings API if we make a new ArrowLabel shape:

- if you have nothing selected, then pointerdown the arrow and drag, the arrow moves and carries the label with it
- if you have nothing selected, then pointerdown the label and drag, the label moves on its own
- if you 'select' both the arrow and the label with a brush selection, pointerdown and drag on either the arrow or the label drags both arrow and label together.

But yeah this isn't possible yet and it shouldn't prevent us from implementing a version of this. We can at least fix the label dragging by doing some clever stuff with the handles infra.

File: packages/tlschema/src/misc/TLHandle.ts [hunk 1]
```diff
@@ -38,4 +40,6 @@ export const handleValidator: T.Validator<TLHandle> = T.object({
 	index: T.string,
 	x: T.number,
 	y: T.number,
+	w: T.optional(T.number),
+	h: T.optional(T.number),
```
[the remark this answers is not in the document]
Comment (mimecuvalo) [reply packages/tlschema/src/misc/TLHandle.ts:44]: I think I might need some hand-holding on what you mean here — I'm not _quite_ sure I follow on what you mean by splitting the TLHandle into two bits. Tap me on the shoulder and we can look at this together?

**Review by SomeHats — COMMENTED** [review]
i wrote these inline comments yesterday before i started playing around on my own branch - i think we discussed most in person, but posting for visibility anyway

**Comment by mimecuvalo** [comment]
Ok! So we've made enough progress to make a first pass at this and we've put it behind a feature flag.

In a followup PR we'll look at:
- @ds300 feedback on making the handle interactive on first click, not after selected in general
- splitting up the `TLHandle` (if that's even still the right place for the interaction)
- make labels offset go to the end of a segment, if there's no arrowhead present

@SomeHats can you take one last look and make sure we're good to land what we have? thank you!

**Review by SomeHats — APPROVED** [review]
Looking good!

> * splitting up the `TLHandle` (if that's even still the right place for the interaction)
This is probably fine in this specific, but in general we should probably be more careful about pushing changes that change our data model at rest. as we're going to be migrating away from storing handles anyway in the follow-up, i think this is ok!

**Pull request merged**